treewide: mark packages that never built ok broken

Mindavi opened the pull request:
###### Description of changes

This is the first result of work I'm doing towards the goal explained in https://github.com/NixOS/nixpkgs/issues/185049.

This seemed like the safest set to start with, since these never worked to begin with.

I can see some packages that might not have been marked with the right platforms (when they were added to nixpkgs), but I wanted to start with this and also show that those things can be caught with this tool.

###### Things done

- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I'm currently running `maintainers/scripts/rebuild-amount.sh` to see how much rebuilds this will cause, which hopefully will give some insight in which packages are removed (ofborg should help with that too, I guess).

Consider this a bit of an RFC as well. For packages that have not been broken forever, I still want to add some extra info in the commit message, but I wasn't sure how useful that'd be here. If you think it would really be useful to add here I could see what I can do, but I think this should be good enough for these types of packages.

Use the following links to check the build list of the packages:
- (for aarch64-darwin) https://hydra.nixos.org/job/nixpkgs/nixpkgs-unstable-aarch64-darwin/bazel_3.aarch64-darwin
- (for the other 3 platforms) https://hydra.nixos.org/job/nixpkgs/trunk/libstdcxx5.aarch64-linux

Fill in the attr and platform as you see fit in the above links.

Mindavi commented:
Hmm, I think I fat-fingered some of the updated packages, looking into that...

Mindavi commented:
Marking as draft since there's some things in here that have multiple versions of which some are working and some aren't. Pointers on how to find those (ideally without an expensive full eval) would be welcome :smile: ). I guess the alternative would be to do an actual full eval and just filter any out that affect > 1 attribute for manual consideration.

SuperSandro2000 commented on pkgs/games/openspades/default.nix:
```diff
@@ -58,7 +58,7 @@ stdenv.mkDerivation rec {
 
   meta = with lib; {
-    broken = stdenv.isDarwin;
+    broken = (stdenv.isDarwin && stdenv.isAarch64) || (stdenv.isLinux && stdenv.isAarch64) || (stdenv.isDarwin && stdenv.isx86_64);  # never built on aarch64-linux since first introduction in nixpkgs
```
for example here we definitely want to split the comment into an extra line

Mindavi replied:
I agree

SuperSandro2000 reviewed (commented):
Thanks for working on this. Some early feedback from me.
Where can I find the code you used to generate this?

Can we please fix an earlier error and move broken to the end of meta?

Also for longer lines I would like to move the commend above.

If both x86 and aarch64 of darwin are marked broken then we can also just admit it and simplify the code to just mark it broken for darwin.

If no or is used then we can ommit the ().

Mindavi committed:
treewide: mark packages broken that never built on PLATFORM

Done with the help of https://github.com/Mindavi/nixpkgs-mark-broken
Tool is still WIP but this is one of the first results.

I manually audited the results and removed some results that were not valid.

Note that some of these packages maybe should have more constrained platforms set
instead of broken set, but I think not being perfectly correct is better than
just keep trying to build all these things and never succeeding.

Some observations:

- Some darwin builds require XCode tools
- aarch64-linux builds sometimes suffer from using gcc9
  - gcc9 is getting older and misses some new libraries/features
- Sometimes tools try to do system detection or expect some explicit settings for
  platforms that are not x86_64-linux

Mindavi commented:
I manually audited the whole list (and as such left out things that affected attrs that were not actually broken) and applied all suggestions from @SuperSandro2000 . Care to re-review?

Source code for the tool can be found here: https://github.com/Mindavi/nixpkgs-mark-broken/. It's still pretty rough, but hopefully gives some insights regardless.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix:
```diff
@@ -229,5 +228,7 @@ stdenv.mkDerivation rec {
     changelog = "https://github.com/kotatogram/kotatogram-desktop/releases/tag/k{version}";
     maintainers = with maintainers; [ ilya-fedin ];
+    # never built on aarch64-darwin since first introduction in nixpkgs
+    broken = (stdenv.isDarwin && stdenv.isAarch64) || (stdenv.isLinux && stdenv.isAarch64);
```
Not sure if we would want to use stdenv.isAarch64 here only or not

Mindavi replied:
My gut feeling says that's not a good idea, but it can be considered.

Mindavi commented:
I'll fix the blesh issue, but don't understand why it happened since my script tries to eval after setting and aborts if it doesn't. Or at least that is the intention.

Mindavi commented:
I think this is ready to go, let me know if there's anything else you'd like to see here. I'll look further into why the missing of the `stdenv` variable as input was not caught by my scripts, I am still not very sure about that. Used nix-eval-jobs to find the remaining issues after struggling with finding them for a bit.

Mindavi commented on pkgs/servers/isso/default.nix:
```diff
@@ -63,5 +63,7 @@ with python3Packages; buildPythonApplication rec {
     homepage = "https://posativ.org/isso/";
     license = licenses.mit;
     maintainers = with maintainers; [ fgaz ];
+    # never built on aarch64-darwin since first introduction in nixpkgs
```
Remove, eval fail? I think stdenv is not in this scope, but I guess it is a bit weird with nodePackages

Mindavi replied:
No, this is not an eval fail. I just missed that stdenv was inherited and was distracted by isso being marked as 'changed' by ofborg.

Mindavi commented:
Here goes nothing...

Mindavi commented:
Way too many aarch64-darwin packages seem removed: https://hydra.nixos.org/eval/1787501?full=1#tabs-removed. I don't know if it's due to this, but maybe?

Mindavi commented:
I think subunit was just fixed: https://hydra.nixos.org/job/nixpkgs/nixpkgs-unstable-aarch64-darwin/subunit.aarch64-darwin

https://github.com/NixOS/nixpkgs/pull/201890

Still not sure why the depending builds seemed to be ok but the dependency seemingly not?

Mindavi commented:
Oh interesting, I think the subunit python package inherits some properties from the other derivation, including meta. This causes the broken mark to propagate to the package too...

pkgs/development/python-modules/subunit/default.nix

I'll revert the subunit broken mark later, it can be solved in another way in the future.
